addons/account/models/account_account.py

```diff
@@ -858,6 +957,310 @@ def get_import_templates(self):
+                account_2=account_to_merge_into.with_company(account_to_merge_into.company_ids[:1]).display_name,
+                companies_2=",".join(account_to_merge_into.company_ids.mapped('name')),
+            )
+        if is_irreversible:
+            msg += _(
+                "This cannot be undone because you are merging accounts belonging to the same company.\n"
+                "After merging, we won't be able to separate journal items based on which account they originally referenced."
+            )
+        action = self.env['ir.actions.actions']._for_xml_id('account.action_merge_accounts')
+        raise RedirectWarning(msg, action, _("Merge"), additional_context={**self.env.context, 'account_merge_confirm': True})
+
+    def _action_merge(self, account_to_merge_into):
+        """ Perform the merge of the accounts in `self` into `account_to_merge_into`.
+        This will update the account codes of `account_to_merge_into` based on those of `self`,
+        and update keys in DB from `self` to `account_to_merge_into`.
+        This method expects checks to already have been performed.
+        """
+        # Step 1: Keep track of the company_ids we should write on the account.
+        # We will do so only at the end, to avoid triggering the constraint that prevents duplicate codes.
+        # Writing the codes will be handled by the update of ir_property.
+        company_ids_to_write = self.company_ids
+
+        # Step 2: Check that we have write access to all the accounts and access to all the companies
+        # of these accounts.
+        self.check_access_rights('write')
+        self.check_access_rule('write')
+        if forbidden_companies := (self.sudo().company_ids - self.env.user.company_ids):
+            raise UserError(_(
+                "You do not have the right to perform this operation as you do not have access to the following companies: %s.",
+                ", ".join(c.name for c in forbidden_companies)
+            ))
+
+        # Step 3: Update foreign keys in DB.
+
+        # The dict that contains instructions to update the DB. Its structure is:
+        # {
+        #     'model_name': [
+        #         (field, records_to_update, query)
+        #     ], ...
+        # }
+        accounts_to_remove = self - account_to_merge_into
+        operations = defaultdict(list)
+
+        # 3.1: Add account.account many2many fields, except company_ids.
+        many2many_fields = self.env['ir.model.fields'].search([
+            ('model', '=', 'account.account'),
+            ('ttype', '=', 'many2many'),
+            ('store', '=', True),
+            ('name', '!=', 'company_ids')
+        ])
+        for field_to_update in many2many_fields:
+            query = SQL(
+                """
+                UPDATE %(table)s
+                SET %(column)s = %(account_id_to_merge_into)s
+                WHERE %(column)s IN %(account_ids_to_remove)s
```

Comment:
Indent queries to be consistent with the "new" style convention
```suggestion
                UPDATE %(table)s
                   SET %(column)s = %(account_id_to_merge_into)s
                 WHERE %(column)s IN %(account_ids_to_remove)s
```